Girder 2.2

brianhelba opened the pull request:
Addresses some of #382

brianhelba commented:
@msmolens I need to do more testing on this myself, but feel free to start looking.

brianhelba commented:
The commit for `Displaying modal dialogs in sequence now works natively` doesn't seem to actually work for me.

msmolens commented:
@brianhelba Yes, I still see the dialog issue in some cases, but not always, so there's likely a race condition. I agree that we should drop that commit for now.

msmolens commented on web_external/Datasets/CreateDatasetView.js:
```diff
@@ -179,8 +160,8 @@ const CreateDatasetView = View.extend({
-    uploadFinished: function (files) {
-        this.uploadedZipFiles = _.pluck(files.files, 'name');
+    uploadFinished: function (info) {
+        this.uploadedFiles = info.files;
```
I'm seeing behavior where the File object in `info.files` loses its `id` attribute. I think it's safer to explicitly save the attributes we need, like:

````
this.uploadedFiles = _.map(info.files, (file) => ({id: file.id, name: file.name}));
````

brianhelba replied:
Sounds good. It's probably not a good idea to trust that strange system-created objects will be immutable.

msmolens commented on web_external/models/UserModel.js:
```diff
@@ -33,39 +34,53 @@ UserModel.prototype.canAcceptTerms = function () {
+    .fail((resp) => {
+        deferred.reject(resp);
     });
+    return deferred.promise();
```
I think we should be able to avoid this pattern of creating and resolving/rejecting a new Deferred. Something like:
````
    return restRequest({
        path: 'user/acceptTerms',
        type: 'POST'
    })
    .done((resp) => {
        if (_.has(resp, 'extra') && resp.extra === 'hasPermission') {
            // Directly update user permissions
            this.get('permissions').acceptTerms = true;
            this.trigger('change:permissions');
    })
````

brianhelba replied:
I mostly agree. We'll still have to create a new Deferred here, but we can do it within the `.then`. Ideally, we should just be able to use `return` and `throw` within `.then`, but jQuery 2 doesn't support that yet.

msmolens commented on web_external/models/UserModel.js:
```diff
@@ -33,39 +34,53 @@ UserModel.prototype.canAcceptTerms = function () {
+    .fail((resp) => {
+        deferred.reject(resp);
     });
+    return deferred.promise();
```
Here too. Maybe:
````
    return restRequest({
        path: 'user/requestCreateDatasetPermission',
        type: 'POST'
    })
    .then((resp) => {
        if (_.has(resp, 'extra') && resp.extra === 'hasPermission') {
            // Directly update user permissions
            this.get('permissions').createDataset = true;
            this.trigger('change:permissions');
        } else {
            return $.Deferred().reject(resp);
        }
    })
````

brianhelba replied:
Totally agreed. jQuery 2 is a little weird in how it handles `.then`, so I wanted to be safe, but just returning a thenable should still work.

brianhelba commented:
@msmolens Really good feedback, thanks. I'l fix these all.

brianhelba commented:
@msmolens Ok, I've dropped the `Displaying modal dialogs in sequence now works natively` commit from the middle and added two new commits at the end to address your feedback.

PTAL.

brianhelba commented:
@msmolens Full disclosure, I haven't had a chance to test these last two changes. I'll do so later if you don't get to it first.
